nullaway/src/main/java/com/uber/nullaway/fixserialization/Serializer.java

```diff
@@ -213,9 +217,45 @@ public static String serializeSymbol(@Nullable Symbol symbol) {
         return symbol.name.toString();
       case METHOD:
       case CONSTRUCTOR:
-        return symbol.toString();
+        return serializeMethodSignature((Symbol.MethodSymbol) symbol);
       default:
         return symbol.flatName().toString();
     }
   }
+
+  /**
+   * Serializes the signature of the given {@link Symbol.MethodSymbol} to a string. This code is
+   * based on {@link com.google.errorprone.util.Signatures#prettyMethodSignature}}.
+   *
+   * @param methodSymbol The method symbol to serialize.
+   * @return The serialized method symbol.
+   */
+  private static String serializeMethodSignature(Symbol.MethodSymbol methodSymbol) {
+    StringBuilder sb = new StringBuilder();
+    if (methodSymbol.isConstructor()) {
+      Symbol.ClassSymbol encClass = methodSymbol.owner.enclClass();
+      // For constructors, we use the name of the enclosing class.
+      Name name = encClass.getSimpleName();
+      if (name.isEmpty()) {
+        // For anonymous classes, we use the name of the superclass.
```

Comment:
Fair enough, since anonymous classes can't declare constructors, so if we see an error for:

```
List list = new ArrayList(10) {...}
```

the constructor involved will literally have to be `ArrayList(int)` (though the class won't match, since that constructor is declared in `java.util.ArrayList` and our `encClass` will be `com.example.Foo$1` or similar...).

Actually, meta note here: we have a big number of corner cases here to handle serializing constructors. Do we have existing test cases in `NullAwaySerializationTest.java` for:

a) `ErrorDisplay` with type/kind `METHOD` and a standard constructor
b) This case for anonymous inner class "constructors".

?

(Basically, do we have coverage here for both branches of the conditional above?)

Reply:
@lazaroclapp's comment made me think of something else.  What uniqueness property do we need for these method signatures?  If we have two anonymous subclasses of the same superclass, the signatures will be identical.  And, both of these anonymous classes could appear in the same containing method.  Is that a problem?  Or are we guaranteed that other information alongside the method signature (e.g., a source location for the error) will give us the uniqueness property we need?  It'd be good to document somewhere the required uniqueness properties.

Reply:
We are serializing the enclosing class, so we would get method `ArrayList(int)` in `Foo$1` and method `ArrayList(int)` in `Foo$2`. The suppression would end up in the method containing both anonymous inner classes, but that's just because there is nowhere to put it on the classes themselves, not because we don't have that info. At least that's my understanding. Is this correct, @nimakarimipour ?

Reply:
@msridhar @lazaroclapp Yes the enclosing class is always serialized along the method signature and the combination of two makes the pair unique. Regarding the conditions in the beginning of this method, I was just following logic implemented in `ErrorProne` rather than calling the `getName()` on the method symbol. But after trying just calling `getName()` on the symbol, realized that that simple name for constructors is always `<init>` and not the class name, therefore, we have to manually locate the enclosing class for constructors. Regarding the case where the constructor is inside an anonymous class, I don't have an example as of now that how it can happen to test it. But I think we can just keep this branch to be cautious that if for any reason the execution hit that branch, it serialize something rather than empty. Please let me know your thoughts on this. I updated comments and added a test case specifically for constructors. 
76c28d617befba6ee3f5d4508caaef282f097a24

Reply:
An anonymous class _cannot_ declare its own constructors, period.  I believe the only case where `name.isEmpty()` would return `true` is when you have a constructor _call_ for an new anonymous class, i.e., something like `new ArrayList(10) {...}`.  For this `NewClassTree`, I think if you look at the `MethodSymbol` for the invoked method, it will be a constructor where the name of the enclosing class is empty.  But, for our use case, we should never be trying to serialize such a `MethodSymbol`, correct @nimakarimipour?  If that is the case, I would rather throw a `RuntimeException` here rather than using the superclass name.  Then, we will know if we were correct in our assumption that this is impossible, or whether there is a case that we missed and need to think about more.

Reply:
@msridhar Yes, I agree and saw just NullAway logic for handling `matchNewClass` in this case. Changed it throw an exception 341c950fd431523ca6ddd86d9f4bf6d0b6b945a5.

Reply:
Also added a comment ddb603c6595e2eeab3061ad182978220be2b98f5